### DISCUSSION TITLE
Unicode Characters

### DISCUSSION THREAD

**Discussion by ThuanLebk** [mike-lischke/vscode-antlr4#208] [Q&A] [answered]
I have problem when generating sentence that contains so many unreadable Unicode characters. How to restrict allowed characters to just ASCII?
Thank you very much.

**Comment by mike-lischke** [comment, marked as answer]
There's no way to influence the set used for printable Unicode characters. So you have only 2 options:

1. Change the grammar temporarily to use only the ASCII script for ID parsing.
2. Specify a number of fixed IDs as described in the [Rule Mappings](https://github.com/mike-lischke/vscode-antlr4/blob/master/doc/sentence-generation.md#rule-mappings) documentation.

**Reply by ThuanLebk** [reply]
Thanks for the reply.
I have some questions related given this simple grammar:
##############
grammar test2;
program: statment EOF;
statment: ids '=' exps;
exps: exp;
exp: exp '+' term | term;
term: term '*' op | op;
op: Int | '(' exps ')';
ids: Id (',' Id)*;
ints: Int (',' Int)*;
Id: [a-z]+;
Int: [0-9]+;
WS: [ \t\r\n] -> skip;
ERROR_CHAR: .;
################
The problem comes from 'op' rule, specifically with '(' exps ')'. (Replacing exps with Int would be fine).
Example output: 2) o  = 0  * ( ⨱ ) 
Many other output has character ⨱. Could you suggest some reasons why it was that printing that character? I suppose it has something to do with recursion.
